usr/src/man/man1/crle.1
'\" te
.\"  Copyright (c) 2008, Sun Microsystems, Inc. All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH CRLE 1 "Aug 19, 2022"
.SH NAME
crle \- configure runtime linking environment
.SH SYNOPSIS
.nf
\fBcrle\fR [\fB-64\fR] [\fB-a\fR \fIname\fR] [\fB-A\fR \fIname\fR] [\fB-c\fR \fIconf\fR] [\fB-e\fR \fIenv\fR] [\fB-E\fR \fIenv\fR]
     [\fB-f\fR \fIflags\fR] [\fB-i\fR \fIname\fR] [\fB-I\fR \fIname\fR] [\fB-g\fR \fIname\fR] [\fB-G\fR \fIname\fR]
     [\fB-l\fR \fIdir\fR] [\fB-o\fR \fIdir\fR] [\fB-s\fR \fIdir\fR] [\fB-u\fR] [\fB-v\fR]
.fi

.SH DESCRIPTION
The \fBcrle\fR utility provides for the creation and display of a runtime
linking configuration file. The configuration file is read and interpreted by
the runtime linker, \fBld.so.1\fR(1), during process startup. The runtime
linker attempts to read a default configuration file for all processes. For
32-bit processes, the default configuration file is \fB/var/ld/ld.config\fR.
For 64-bit processes, the default configuration file is
\fB/var/ld/64/ld.config\fR.
.sp
.LP
Without any arguments, or with just the \fB-c\fR option, \fBcrle\fR displays
configuration information. This information includes the contents of a
configuration file, any system defaults and the command-line required to
regenerate the configuration file. When used with any other options, a new
configuration file is created or updated.
.sp
.LP
The runtime linker can also be directed to an alternative configuration file by
setting one of the \fBLD_CONFIG\fR family of environment variable.
\fBLD_CONFIG\fR applies to both 32-bit and 64-bit programs. Since 32-bit and
64-bit configuration files differ, a single configuration file cannot be used
for both class of object. Hence, \fBLD_CONFIG\fR can adversely affect program
execution in cases where a program of one class executes a program of the other
class. In particular, it is common practice for the 32-bit version of standard
Solaris utilities to execute their 64-bit counterpart. \fBLD_CONFIG\fR cannot
be successfully used in this case. Therefore, the use of the \fBLD_CONFIG_32\fR
and \fBLD_CONFIG_64\fR environment variables, that precisely target the
appropriate class of process, is recommended.
.sp
.LP
Creating an incorrect configuration file in the standard location,
\fB/var/ld\fR, can prevent programs from running, and can therefore be
difficult to recover from. To guard against this situation, it is recommended
that new configuration files first be created in a temporary location. Then set
the appropriate \fBLD_CONFIG\fR environment variable to this new configuration
file. This setting causes the new configuration file to be used by the runtime
linker instead of any default. After verification, the new configuration file
can be moved to the default location if desired. At any time, the environment
variable \fBLD_NOCONFIG\fR can be set to any value to instruct the runtime
linker to ignore any configuration files. This setting can prove useful during
experimentation.
.sp
.LP
A configuration file can contain the following information.
.sp
.ne 2
.na
\fBDefault Search Paths\fR
.ad
.sp .6
.RS 4n
The runtime linker uses a prescribed search path for locating the dynamic
dependencies of an object. This search path starts with the components of any
\fBLD_LIBRARY_PATH\fR definition, followed by the components of an object's
\fBrunpath\fR. Finally, any default search paths specific to the object's class
are used. This last component of the search path can be expressed within the
configuration file. Typically, use of this facility should be augmented with
any system default. See the \fB-l\fR and \fB-u\fR options.
.RE

.sp
.ne 2
.na
\fBTrusted Directories\fR
.ad
.sp .6
.RS 4n
When processing a secure application, the runtime linker restricts the use of
\fBLD_LIBRARY_PATH\fR searches, and \fB$ORIGIN\fR token expansion. See
\fISecurity\fR in \fILinker and Libraries Guide\fR. In addition, the
directories from which preload and audit libraries can be located are also
restricted. The path names that are associated with preload and audit libraries
are restricted to known trusted directories. Trusted directories can be
expressed within the configuration file. Typically, use of this facility should
be augmented with any system defaults. See the \fB-s\fR and \fB-u\fR options.
.RE

.sp
.ne 2
.na
\fBEnvironment Variables\fR
.ad
.sp .6
.RS 4n
Any environment variable interpreted by the runtime linker can be specified
within the configuration file.
.RE

.sp
.ne 2
.na
\fBDirectory Cache\fR
.ad
.sp .6
.RS 4n
The location of shared objects within defined directories can be maintained as
a cache within the configuration file. This directory cache can reduce the
overhead of searching for application dependencies.
.RE

.sp
.ne 2
.na
\fBAlternative Objects\fR
.ad
.sp .6
.RS 4n
In conjunction with the directory cache, shared objects can have alternative
objects specified for use at runtime. These alternate objects, can be supplied
by the user. Alternative objects can also be created by \fBcrle\fR as copies of
shared objects fixed to known memory locations. These fixed alternative objects
can require less processing at runtime than their original shared object
counterpart.
.RE

.sp
.LP
Defining additional default search paths, or additional trusted directories can
be useful for administrators who wish to install third party software in a
central location, or otherwise alter the search path of applications that might
not have been coded with a suitable runpath.
.sp
.LP
The declaration of alternative objects provides a means of replacing
dependencies other than by using symbolic links or requiring
\fBLD_LIBRARY_PATH\fR settings.
.sp
.LP
The declaration of environment variables that are interpreted by the runtime
linker provides a means of centralizing their definition for all applications.
.sp
.LP
The directory cache, and \fBcrle\fR generated alternate objects, can provide a
means of reducing the runtime startup overhead of applications. Alternative
objects can be useful for applications that require many dependencies, or whose
dependencies are expensive to relocate. Shared objects that contain
\fBposition-dependent\fR code are often expensive to relocate. Note, the system
has many caching facilities that help mitigate expenses such as negative path
lookups, and thus employing \fBcrle\fR to create a directory cache may have
minimal effect other than for some very specific cases.
.sp
.LP
When alternate objects that are generated by \fBcrle\fR are specified within a
configuration file, the runtime linker performs some minimal consistency
verification. The alternative objects are verified against their originating
objects. This verification is intended to avert application failure should an
applications configuration information become out-of-sync with the underlying
system components. When this situation arises the flexibility offered by
dynamic linking system components can be compromised. This type of application
failure can be very difficult to diagnose. No verification of directory cache
information is performed. Any changes to the directory structure are not seen
by a process until the cache is rebuilt.
.sp
.LP
System shared objects are often well tuned, and can show little benefit from
being cached. The directory cache and alternative object features are typically
applicable to user applications and shared objects, and may only show
improvement in some very specific cases.
.sp
.LP
\fBcrle\fR creates alternate objects for the shared objects that are discovered
when using the \fB-I\fR and \fB-G\fR options, using \fBdldump\fR(3C). The
alternate object is created in the directory specified by the preceding
\fB-o\fR option, or defaults to the directory in which the configuration file
is created. The flags used by \fBdldump()\fR are specified using the \fB-f\fR
option, or default to \fBRTLD_REL_RELATIVE\fR.
.SH OPTIONS
The following options are supported.
.sp
.ne 2
.na
\fB\fB-64\fR\fR
.ad
.sp .6
.RS 4n
Specify to process 64-bit objects, the default is 32-bit. Use \fB-64\fR to
create a 64-bit specific configuration file.
.RE

.sp
.ne 2
.na
\fB\fB-a\fR \fIname\fR\fR
.ad
.sp .6
.RS 4n
Create an alternative path name for \fIname\fR. The alternative path name is
added to the configuration file.
.sp
The actual alternative file must be supplied by the user. Multiple occurrences
of this option are permitted. If \fIname\fR is a directory, each shared object
within the directory is added to the cache. If \fIname\fR does not exist, then
\fIname\fR is marked in the cache as a nonexistent file.
.sp
Typically, this option is used with the \fB-o\fR option.
.RE

.sp
.ne 2
.na
\fB\fB-A\fR \fIname\fR\fR
.ad
.sp .6
.RS 4n
Create an optional alternative path name for \fIname\fR. This alternative path
name is added to the configuration file.
.sp
This option mimics the \fB-a\fR option, except that if the alternative is
unavailable at runtime, the original object \fIname\fR is used. This model
mimics the use of auxiliary filters. See \fIGenerating Auxiliary Filters\fR in
\fILinker and Libraries Guide\fR.
.sp
Typically, this option is used with the \fB-o\fR option.
.RE

.sp
.ne 2
.na
\fB\fB-c\fR \fIconf\fR\fR
.ad
.sp .6
.RS 4n
Specify to use the configuration file name \fIconf\fR. If this option is not
supplied, the default configuration file is used.
.RE

.sp
.ne 2
.na
\fB\fB-e\fR \fIenv\fR\fR
.ad
.sp .6
.RS 4n
Specify a \fBreplaceable\fR environment variable, \fIenv\fR. Only environment
variables that are applicable to the runtime linker are meaningful. Multiple
occurrences of this option are permitted. This option is similar to the
\fB-E\fR option. However, the options differs in how configuration file
definitions, and process environment definitions of the same name are resolved
at runtime.
.sp
A definition established in a configuration file can be \fBoverridden\fR by a
process environment definition, or be \fBsuppressed\fR by a null-value process
environment definition.
.sp
In other words, these configuration file definitions can be replaced, or
removed by the process environment at runtime.
.RE

.sp
.ne 2
.na
\fB\fB-E\fR \fIenv\fR\fR
.ad
.sp .6
.RS 4n
Specify a \fBpermanent\fR environment variable, \fIenv\fR. Only environment
variables that are applicable to the runtime linker are meaningful. Multiple
occurrences of this option are permitted. This option is similar to the
\fB-e\fR option. However, the option differs in how configuration file
definitions, and process environment definitions of the same name are resolved
at runtime.
.sp
Environment variable definitions that are meaningful to the runtime linker fall
into one of two categories. Singular definitions are definitions such as
\fBLD_NOLAZYLOAD=1\fR and \fBLD_DEBUG_OUTPUT=\fR\fIfile\fR. List definitions,
which can take one or more values, are definitions such as
\fBLD_LIBRARY_PATH=\fR\fIpath\fR, and \fBLD_DEBUG=\fR\fIfiles\fR,\fIdetails\fR.
.sp
A singular definition that is established in a configuration file takes
precedence over a process environment definition. A list definition that is
established in a configuration file is \fBappended\fR to a process environment
definition. Any definition that is established in a configuration file can
\fBnot\fR be suppressed by a null-value process environment definition.
.sp
In other words, these configuration file definitions can \fBnot\fR be replaced,
or removed by the process environment at runtime.
.RE

.sp
.ne 2
.na
\fB\fB-f\fR \fIflags\fR\fR
.ad
.sp .6
.RS 4n
Provide the symbolic \fIflags\fR argument to the \fBdldump\fR(3C) calls used to
generate alternate objects. Any of the \fBRTLD_REL\fR flags that are defined in
\fB/usr/include/dlfcn.h\fR can be used. Multiple flags can be \fBor\fR'ed
together using the "\fB|\fR" character. In this case, the string should be
quoted to avoid expansion by the shell. If no \fIflags\fR values are provided
the default flag is \fBRTLD_REL_RELATIVE\fR.
.RE

.sp
.ne 2
.na
\fB\fB-i\fR \fIname\fR\fR
.ad
.sp .6
.RS 4n
Add an individual \fIname\fR to the configuration cache. Multiple occurrences
of this option are permitted. \fIname\fR can be a shared object or a directory.
If \fIname\fR is a directory, each shared object within the directory is added
to the cache. If \fIname\fR does not exist, the \fIname\fR is marked in the
cache as a nonexistent directory.
.RE

.sp
.ne 2
.na
\fB\fB-I\fR \fIname\fR\fR
.ad
.sp .6
.RS 4n
Mimic the \fB-i\fR, and in addition any shared object that is processed has an
alternative created using \fBdldump\fR(3C). If the \fB-f\fR flag contains
\fBRTLD_REL_EXEC\fR, then \fIname\fR can be a dynamic executable, for which an
alternative is created. Only one dynamic executable can be specified in this
manner, as the cache that is created is specific to this application.
.RE

.sp
.ne 2
.na
\fB\fB-g\fR \fIname\fR\fR
.ad
.sp .6
.RS 4n
Add the group \fIname\fR to the configuration cache. Each object is expanded to
determine its dependencies. Multiple occurrences of this option are permitted.
\fIname\fR can be a dynamic executable, shared object or a directory. If
\fIname\fR is a shared object, the shared object and its dependencies are added
to the cache. If \fIname\fR is a directory, each shared object within the
directory, and its dependencies, are added to the cache.
.RE

.sp
.ne 2
.na
\fB\fB-G\fR \fIname\fR\fR
.ad
.sp .6
.RS 4n
Mimic the \fB-g\fR option, and in addition any shared object that is processed
has an alternative created using \fBdldump\fR(3C). If \fIname\fR is a dynamic
executable, and the \fB-f\fR flag contains \fBRTLD_REL_EXEC\fR, then an
alternative for the dynamic executable is also created. Only one dynamic
executable can be specified in this manner as the cache that is created is
specific to this application.
.RE

.sp
.ne 2
.na
\fB\fB-l\fR \fIdir\fR\fR
.ad
.sp .6
.RS 4n
Specify a new default search directory \fIdir\fR for \fBELF\fR
objects. Multiple occurrences of this option are permitted.
.sp
The default search paths for 32-bit \fBELF\fR objects are \fB/lib\fR followed
by \fB/usr/lib\fR. For 64-bit \fBELF\fR objects, the default search paths are
\fB/lib/64\fR followed by \fB/usr/lib/64\fR.
.sp
Use of this option \fBreplaces\fR the default search path. Therefore, a
\fB-l\fR option is normally required to specify the original system default in
relation to any new paths that are being applied. However, if the \fB-u\fR
option is in effect, and a configuration file does \fBnot\fR exist, the system
defaults are added to the new configuration file. These defaults are added
before the new paths specified with the \fB-l\fR option.
.RE

.sp
.ne 2
.na
\fB\fB-o\fR \fIdir\fR\fR
.ad
.sp .6
.RS 4n
When used with either the \fB-a\fR or \fB-A\fR options, specifies the directory
\fIdir\fR in which any alternate objects exist. When alternative objects are
created by \fBcrle\fR, this option specified where the alternative are created.
Without this option, alternate objects exist in the directory in which the
configuration file is created. Multiple occurrences of this option are
permitted, the directory \fIdir\fR being used to locate alternatives for any
following command-line options. Alternative objects are not permitted to
override their associated originals.
.sp
Typically, this option is used with the \fB-a\fR or \fB-A\fR options.
.RE

.sp
.ne 2
.na
\fB\fB-s\fR \fIdir\fR\fR
.ad
.sp .6
.RS 4n
Specify a new trusted directory \fIdir\fR for \fIsecure\fR \fBELF\fR
objects. See \fBSECURITY\fR in \fBld.so.1\fR(1) for a definition of secure
objects. See \fISecurity\fR in \fILinker and Libraries Guide\fR for a
discussion of runtime restrictions imposed on secure applications.
.sp
Multiple occurrences of this option are permitted.
.sp
The default trusted directories for secure 32-bit \fBELF\fR objects is
\fB/lib/secure\fR followed by \fB/usr/lib/secure\fR.  For 64-bit secure
\fBELF\fR objects, the default trusted directories are \fB/lib/secure/64\fR
followed by \fB/usr/lib/secure/64\fR.
.sp

usr/src/cmd/sgs/crle/common/crle.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<fcntl.h>
#include	<stdio.h>
#include	<string.h>
#include	<unistd.h>
#include	<locale.h>
#include	<dlfcn.h>
#include	<errno.h>
#include	"_crle.h"
#include	"conv.h"
#include	"msg.h"


/*
 * crle(1) entry point and argument processing.
 *
 * Two passes of the arguments are carried out; the first collects any single
 * instance options and establishes defaults that might be appropriate for
 * other arguments:
 *
 *  -64		operate on, or apply, 64-bit objects (default is 32-bit).
 *
 *  -c file	defines the output configuration file.
 *
 *  -f flag	flags for dldump(3dl).
 *
 *  -o dir	defines the output directory for any dldump(3dl) objects
 *		that follow.  For backward compatibility (RTC_VER_ONE only
 *		allowed one output directory) allow the first occurrence of this
 *		specification to catch any previous files.  If not specified,
 *		the configuration files parent directory is used).
 *
 *  -u		update any existing configuration file.  Any additional
 *		arguments supplied will be added to the new configuration
 *		information.
 *
 *  -v		verbose mode.
 *
 * The second pass collects all other options and constructs an internal
 * string table which will be used to create the eventual configuration file.
 *
 *  -a name	add the individual name, with an alternative to the
 *		configuration cache.  No alternative is created via dldump(3dl),
 *		it is the users responsibility to furnish the alternative.
 *
 *  -A name	add the individual name, with an optional alternative to the
 *		configuration cache.  No alternative is created via dldump(3dl),
 *		it is the users responsibility to furnish the alternative.
 *
 *  -e envar	replaceable environment variable
 *
 *  -E envar	permanent environment variable
 *
 *  -i name	add the individual name to the configuration cache.  If name
 *		is a directory each shared object within the directory is added
 *		to the cache.
 *
 *  -I name	same as -i, but in addition any ELF objects are dldump(3dl)'ed.
 *
 *  -g name	add the group name to the configuration cache.  Each object is
 *		expanded to determine its dependencies and these are added to
 *		the cache.  If name is a directory each shared object within the
 *		directory and its dependencies are added to the cache.
 *
 *  -G app	same as -g, but in addition any ELF objects are dldump(3dl)'ed.
 *
 *  -l dir	library search directory
 *
 *  -s dir	trusted (secure) directory
 */

/*
 * Establish a structure for maintaining current object directory attributes.
 * We wish to validate the access of any object directory that will be written
 * to (dldump(3dl), and thus by maintaining a current object directory and its
 * intended use we can perform this validation later.
 */
typedef struct {
	char	*o_objdir;
	uint_t	o_flags;
} Objdir;

/*ARGSUSED2*/
int
main(int argc, char **argv, char **envp)
{
	Crle_desc	crle = { 0 };
	int		c, error = 0;
	char		**lib;
	Alist		*objdirs = NULL;
	Objdir		*objdir, *iobjdir;
	struct stat	ostatus, nstatus;
	int		c_class;

	if ((objdir = iobjdir = alist_append(&objdirs, NULL, sizeof (Objdir),
	    AL_CNT_CRLE)) == NULL)
		return (1);

	/*
	 * Establish locale.
	 */
	(void) setlocale(LC_MESSAGES, MSG_ORIG(MSG_STR_EMPTY));
	(void) textdomain(MSG_ORIG(MSG_SUNW_OST_SGS));

	/*
	 * Initialization configuration information.
	 */
	crle.c_name = argv[0];
	crle.c_flags |= CRLE_ADDID;
	crle.c_strbkts = 503;
	crle.c_inobkts = 251;
	c_class = M_CLASS;

	/*
	 * First argument pass.
	 */
	while ((c = getopt(argc, argv, MSG_ORIG(MSG_ARG_OPTIONS))) != -1) {
		switch (c) {

		case '6':			/* operate on 64-bit objects */
			if (optarg[0] != '4') {
				(void) fprintf(stderr,
				    MSG_INTL(MSG_ARG_ILLEGAL), crle.c_name,
				    MSG_ORIG(MSG_ARG_6), optarg);
				error = 1;
			}

			c_class = ELFCLASS64;
			break;

		case 'A':			/* create optional */
			/* FALLTHROUGH */	/*	alternative */
		case 'a':			/* create alternative */
			crle.c_flags |= (CRLE_CREAT | CRLE_ALTER);
			objdir->o_flags |= (CRLE_CREAT | CRLE_ALTER);
			break;

		case 'c':			/* define the config file */
			if (crle.c_confil) {
				(void) fprintf(stderr, MSG_INTL(MSG_ARG_MULT),
				    crle.c_name, MSG_ORIG(MSG_ARG_C));
				error = 1;
			}
			crle.c_confil = optarg;
			break;

		case 'e':			/* replaceable env variable */
			crle.c_flags |= (CRLE_RPLENV | CRLE_CREAT);
			break;

		case 'E':			/* permanent env variable */
			crle.c_flags |= (CRLE_PRMENV | CRLE_CREAT);
			break;

		case 'f':			/* dldump(3dl) flags */
			if (crle.c_dlflags) {
				(void) fprintf(stderr, MSG_INTL(MSG_ARG_MULT),
				    crle.c_name, MSG_ORIG(MSG_ARG_F));
				error = 1;
			}
			if ((crle.c_dlflags = dlflags(&crle,
			    (const char *)optarg)) == 0)
				error = 1;
			break;

		case 'G':			/* group object */
			crle.c_flags |= (CRLE_DUMP | CRLE_ALTER);
			objdir->o_flags |= (CRLE_DUMP | CRLE_ALTER);
			/* FALLTHROUGH */
		case 'g':
			crle.c_flags |= CRLE_CREAT;
			objdir->o_flags |= CRLE_CREAT;
			break;

		case 'I':			/* individual object */
			crle.c_flags |= (CRLE_DUMP | CRLE_ALTER);
			objdir->o_flags |= (CRLE_DUMP | CRLE_ALTER);
			/* FALLTHROUGH */
		case 'i':
			crle.c_flags |= CRLE_CREAT;
			objdir->o_flags |= CRLE_CREAT;
			break;

		case 'l':			/* library search path */
			crle.c_flags |= (CRLE_EDLIB | CRLE_CREAT);
			break;

		case 'o':			/* define an object directory */
			if (objdir->o_objdir) {
				if ((objdir = alist_append(&objdirs, NULL,
				    sizeof (Objdir), AL_CNT_CRLE)) == NULL)
					return (1);
			}
			objdir->o_objdir = optarg;
			break;

		case 's':			/* trusted (secure) path */
			crle.c_flags |= (CRLE_ESLIB | CRLE_CREAT);
			break;

		/*
		 * Search path type, undocumented but left for compatibility.
		 * Previously used to select between AOUT and ELF, now
		 * anything other than ELF is an error.
		 */
		case 't':
			if (strcmp((const char *)optarg,
			    MSG_ORIG(MSG_STR_ELF)) != 0) {
				(void) fprintf(stderr, MSG_INTL(MSG_ARG_TYPE),
				    crle.c_name, optarg);
				error = 1;
			}
			break;

		case 'u':			/* update mode */
			crle.c_flags |= (CRLE_CREAT | CRLE_UPDATE);
			break;

		case 'v':			/* verbose mode */
			crle.c_flags |= CRLE_VERBOSE;
			break;

		default:
			error = 2;
		}
	}

	if (optind != argc)
		error = 2;

	/*
	 * Determine the configuration file, which in the case of an existing
	 * error condition is required in the final error message.
	 */
	if (crle.c_confil == NULL) {
		crle.c_flags |= CRLE_CONFDEF;
		if (c_class == ELFCLASS32) {
			crle.c_confil = (char *)MSG_ORIG(MSG_PTH_CONFIG);
		} else {
			crle.c_confil = (char *)MSG_ORIG(MSG_PTH_CONFIG_64);
		}
	}

	/*
	 * Now that we've generated as many file/directory processing errors
	 * as we can, return if any fatal error conditions occurred.
	 */
	if (error) {
		if (error == 2) {
			(void) fprintf(stderr, MSG_INTL(MSG_ARG_USAGE),
			    crle.c_name);
		} else if (crle.c_flags & CRLE_CREAT) {
			(void) fprintf(stderr, MSG_INTL(MSG_GEN_CREATE),
			    crle.c_name, crle.c_confil);
		}
		return (1);
	}

	/*
	 * Apply any additional defaults.
	 */
	if (crle.c_dlflags == 0)
		crle.c_dlflags = RTLD_REL_RELATIVE;

	crle.c_audit = (char *)MSG_ORIG(MSG_ENV_LD_AUDIT);

	(void) elf_version(EV_CURRENT);

	/*
	 * If we're updating an existing file or not creating a configuration
	 * file at all, investigate the original.
	 */
	if ((crle.c_flags & CRLE_UPDATE) ||
	    ((crle.c_flags & CRLE_CREAT) == 0)) {
		switch (inspectconfig(&crle, c_class)) {
		case INSCFG_RET_OK:
			if ((crle.c_flags & CRLE_UPDATE) == 0)
				return (0);
			break;
		case INSCFG_RET_FAIL:
			return (1);
		case INSCFG_RET_NEED64:
			c_class = ELFCLASS64;
			break;
		}
	}

	/*
	 * Ensure that the right version (32 or 64-bit) of this program
	 * is running. The 32 and 64-bit compilers may align fields within
	 * structures differently. Using the right version of crle for
	 * the config file ensures that all linker components will see
	 * the same layout, without the need for special code.
	 */
#ifdef _ELF64
	if (c_class == ELFCLASS32) {
		(void) fprintf(stderr, MSG_INTL(MSG_ARG_CLASS),
		    crle.c_name, crle.c_confil);
		return (1);
	}
#else
	if (c_class == ELFCLASS64) {
		(void) conv_check_native(argv, envp);

		/*
		 * conv_check_native() should not return, as we expect
		 * the 64-bit version to have executed on top of us.
		 * If it does, it means there is no 64-bit support
		 * available on this system.
		 */
		(void) fprintf(stderr, MSG_INTL(MSG_ISA32_NO64SUP),
		    crle.c_name);
		return (1);
	}
#endif

	if (crle.c_flags & CRLE_VERBOSE)
		(void) printf(MSG_INTL(MSG_DIA_CONFILE), crle.c_confil);

	/*
	 * Make sure the configuration file is accessible.  Stat the file to
	 * determine its dev number - this is used to determine whether the
	 * temporary configuration file we're about to build can be renamed or
	 * must be copied to its final destination.
	 */
	(void) umask(022);
	if (access(crle.c_confil, (R_OK | W_OK)) == 0) {
		crle.c_flags |= CRLE_EXISTS;

		if (stat(crle.c_confil, &ostatus) != 0) {
			int err = errno;
			(void) fprintf(stderr, MSG_INTL(MSG_SYS_OPEN),
			    crle.c_name, crle.c_confil, strerror(err));
			return (1);
		}
	} else if (errno != ENOENT) {
		int err = errno;
		(void) fprintf(stderr, MSG_INTL(MSG_SYS_ACCESS), crle.c_name,
		    crle.c_confil, strerror(err));
		return (1);
	} else {
		int	fd;

		/*
		 * Try opening the file now, if it works delete it, there may
		 * be a lot of processing ahead of us, so we'll come back and
		 * create the real thing later.
		 */
		if ((fd = open(crle.c_confil, (O_RDWR | O_CREAT | O_TRUNC),
		    0666)) == -1) {
			int err = errno;
			(void) fprintf(stderr, MSG_INTL(MSG_SYS_OPEN),
			    crle.c_name, crle.c_confil, strerror(err));
			return (1);
		}
		if (fstat(fd, &ostatus) != 0) {
			int err = errno;
			(void) fprintf(stderr, MSG_INTL(MSG_SYS_OPEN),
			    crle.c_name, crle.c_confil, strerror(err));
			return (1);
		}
		(void) close(fd);
		(void) unlink(crle.c_confil);
	}

	/*
	 * If an object directory is required to hold dldump(3dl) output assign
	 * a default if necessary and insure we're able to write there.
	 */
	if (crle.c_flags & CRLE_ALTER) {
		if (objdir->o_objdir == NULL) {
			char	*str;

			/*
			 * Use the configuration files directory.
			 */
			if ((str = strrchr(crle.c_confil, '/')) == NULL)
				objdir->o_objdir =
				    (char *)MSG_ORIG(MSG_DIR_DOT);
			else {
				int	len = str - crle.c_confil;

				if ((objdir->o_objdir =
				    malloc(len + 1)) == NULL) {
					int err = errno;
					(void) fprintf(stderr,
					    MSG_INTL(MSG_SYS_MALLOC),
					    crle.c_name, strerror(err));
					return (1);
				}
				(void) strncpy(objdir->o_objdir,
				    crle.c_confil, len);
				objdir->o_objdir[len] = '\0';
			}
		}

		/*
		 * If we're going to dldump(3dl) images ourself make sure we
		 * can access any directories.
		 */
		if (crle.c_flags & CRLE_DUMP) {
			Objdir	*objdir = NULL;
			Aliste	idx;
			int	err = 0;

			for (ALIST_TRAVERSE(objdirs, idx, objdir)) {
				if (crle.c_flags & CRLE_VERBOSE)
					(void) printf(MSG_INTL(MSG_DIA_OBJDIR),
					    objdir->o_objdir);

				if ((objdir->o_flags & CRLE_DUMP) == 0)
					continue;

				if (access(objdir->o_objdir,
				    (R_OK | W_OK)) != 0) {
					err = errno;
					(void) fprintf(stderr,
					    MSG_INTL(MSG_SYS_ACCESS),
					    crle.c_name, objdir->o_objdir,
					    strerror(err));
				}
			}
			if (err)
				return (1);
		}
	}

	/*
	 * Establish any initial object directory.
	 */
	crle.c_objdir = iobjdir->o_objdir;

	/*
	 * Create a temporary file name in which to build the configuration
	 * information.
	 */
	if ((crle.c_tempname = tempnam(MSG_ORIG(MSG_TMP_DIR),
	    MSG_ORIG(MSG_TMP_PFX))) == NULL) {
		int err = errno;
		(void) fprintf(stderr, MSG_INTL(MSG_SYS_TEMPNAME),
		    crle.c_name, strerror(err));
		return (1);
	}
	if ((crle.c_tempfd = open(crle.c_tempname, (O_RDWR | O_CREAT),
	    0666)) == -1) {
		int err = errno;
		(void) fprintf(stderr, MSG_INTL(MSG_SYS_OPEN),
		    crle.c_name, crle.c_tempname, strerror(err));
		return (1);
	}
	if (stat(crle.c_tempname, &nstatus) != 0) {
		int err = errno;
		(void) fprintf(stderr, MSG_INTL(MSG_SYS_OPEN),
		    crle.c_name, crle.c_tempname, strerror(err));
		return (1);
	}
	if (ostatus.st_dev != nstatus.st_dev)
		crle.c_flags |= CRLE_DIFFDEV;

	/*
	 * Second pass.
	 */
	error = 0;
	optind = 1;
	while ((c = getopt(argc, argv, MSG_ORIG(MSG_ARG_OPTIONS))) != -1) {
		const char	*str;
		int		flag = 0;

		switch (c) {

		case '6':
			break;

		case 'A':			/* alternative is optional */
			flag = RTC_OBJ_OPTINAL;
			/* FALLTHROUGH */
		case 'a':			/* alternative required */
			flag |= (RTC_OBJ_ALTER | RTC_OBJ_CMDLINE);
			if (inspect(&crle, (const char *)optarg, flag) != 0)
				error = 1;
			break;

		case 'c':
			break;

		case 'e':
			if ((flag = addenv(&crle, (const char *)optarg,
			    RTC_ENV_REPLACE)) == 0)
				error = 1;
			else if ((crle.c_flags & CRLE_VERBOSE) && (flag == 1))
				(void) printf(MSG_INTL(MSG_DIA_RPLENV),
				    (const char *)optarg);
			break;

		case 'E':
			if ((flag = addenv(&crle, (const char *)optarg,
			    RTC_ENV_PERMANT)) == 0)
				error = 1;
			else if ((crle.c_flags & CRLE_VERBOSE) && (flag == 1))
				(void) printf(MSG_INTL(MSG_DIA_PRMENV),
				    (const char *)optarg);
			break;

		case 'f':
			break;

		case 'G':			/* group object */
			flag = (RTC_OBJ_DUMP | RTC_OBJ_ALTER);
			/* FALLTHROUGH */
		case 'g':
			flag |= (RTC_OBJ_GROUP | RTC_OBJ_CMDLINE);
			if (inspect(&crle, (const char *)optarg, flag) != 0)
				error = 1;
			break;

		case 'I':			/* individual object */
			flag = (RTC_OBJ_DUMP | RTC_OBJ_ALTER);
			/* FALLTHROUGH */
		case 'i':
			flag |= RTC_OBJ_CMDLINE;
			if (inspect(&crle, (const char *)optarg, flag) != 0)
				error = 1;
			break;

		case 'l':			/* library search path */
			str = MSG_ORIG(MSG_STR_ELF);
			lib = &crle.c_edlibpath;
			if (addlib(&crle, lib, (const char *)optarg) != 0)
				error = 1;
			else if (crle.c_flags & CRLE_VERBOSE)
				(void) printf(MSG_INTL(MSG_DIA_DLIBPTH),
				    str, (const char *)optarg);
			break;

		case 'o':
			crle.c_objdir = optarg;
			break;

		case 's':			/* trusted (secure) path */
			str = MSG_ORIG(MSG_STR_ELF);
			lib = &crle.c_eslibpath;
			if (addlib(&crle, lib, (const char *)optarg) != 0)
				error = 1;
			else if (crle.c_flags & CRLE_VERBOSE)
				(void) printf(MSG_INTL(MSG_DIA_TLIBPTH),
				    str, (const char *)optarg);
			break;

		case 't':
			break;

		case 'u':
			break;

		case 'v':
			break;
		}
	}

	/*
	 * Now that we've generated as many file/directory processing errors
	 * as we can, return if any fatal error conditions occurred.
	 */
	if (error) {
		(void) unlink(crle.c_tempname);
		if (crle.c_flags & CRLE_CREAT) {
			(void) fprintf(stderr, MSG_INTL(MSG_GEN_CREATE),
			    crle.c_name, crle.c_confil);
		}
		return (1);
	}

	/*
	 * Create a temporary configuration file.
	 */
	if (genconfig(&crle) != 0) {
		(void) unlink(crle.c_tempname);
		return (1);
	}

	/*
	 * If dldump(3dl) images are required spawn a process to create them.
	 */
	if (crle.c_flags & CRLE_DUMP) {
		if (dump(&crle) != 0) {
			(void) unlink(crle.c_tempname);
			return (1);
		}
	}

	/*
	 * Copy the finished temporary configuration file to its final home.
	 */
	if (updateconfig(&crle) != 0)
		return (1);

	return (0);
}
